src/stir/prover_state.rs

```diff
@@ -0,0 +1,349 @@
+use ark_crypto_primitives::{
+    merkle_tree::{Config as MerkleConfig, MerkleTree, MultiPath},
+    sponge::{Absorb, CryptographicSponge},
+};
+use ark_ff::{FftField, PrimeField};
+use ark_poly::{univariate::DensePolynomial, DenseUVPolynomial, EvaluationDomain, Polynomial};
+
+#[cfg(not(feature = "std"))]
+use ark_std::{vec, vec::Vec};
+
+use crate::{
+    domain::Domain,
+    poly_utils,
+    utils::{dedup, proof_of_work, squeeze_integer, stack_evaluations},
+};
+
+use super::{config::STIRConfig, proof::STIRProofRound};
+
+pub struct STIRProverState<F, M, S>
+where
+    F: FftField,
+    M: MerkleConfig,
+    S: CryptographicSponge,
+{
+    answer_coeff: DensePolynomial<F>,
+    domain: Domain<F>,
+    challenge_answers: MultiPath<M>,
+    challenge_values: Vec<Vec<F>>,
+    challenges: Vec<usize>,
+    commitment: MerkleTree<M>,
+    committed_values: Vec<Vec<F>>,
+    config: STIRConfig<M, S>,
+    folding_randomness: F,
+    last_round_commitment: MerkleTree<M>,
+    last_round_committed_values: Vec<Vec<F>>,
+    last_round_domain_size: usize,
+    out_of_domain_samples: Vec<F>,
+    out_of_domain_evaluations: Vec<F>,
+    proof_of_work_nonce: Option<usize>,
+    proximity_generator_randomness: F,
+    quotient_answers: Vec<F>,
+    quotient_set: Vec<F>,
+    round_num: usize,
+    shake_coeff: DensePolynomial<F>,
+    sponge: S,
+    witness_coeff: DensePolynomial<F>,
+}
+
+impl<F, M, S> STIRProverState<F, M, S>
+where
+    F: FftField + PrimeField + Absorb,
+    M: MerkleConfig<Leaf = Vec<F>> + Clone,
+    M::InnerDigest: Absorb,
+    S: CryptographicSponge,
+    S::Config: Clone,
+{
+    pub fn new(
+        domain: Domain<F>,
+        commitment: MerkleTree<M>,
+        committed_values: Vec<Vec<F>>,
+        config: STIRConfig<M, S>,
+        witness_coeff: DensePolynomial<F>,
+    ) -> Self {
+        let mut sponge = S::new(&config.sponge_config);
+        sponge.absorb(&commitment.root());
```

Comment:
@mmaker This is my first time looking at Nimue. The pattern I have here is `absorb/ squeeze` using this sponge. It looks v similar to this example in the Nimue docs: https://docs.rs/nimue/latest/nimue/#protocol-transcripts

Basically, if I swap the sponge in this struct for `IOPattern::<H>::new` and use it the same way, does this fix or improve some issue that I'm unaware of?

Reply:
hey, yes pretty much! 
The prover will be a `Merlin` instance, which can be generated from `iopattern.into_merlin()`. You can either have a mutable reference to it or just store the iopattern inside the config.

Note that if you have a similar struct for the verifier you'd have instead to use`Arthur`, which can generated from (bytes of) the proof and the iopattern, but will keep *a reference* to the proof.